src/main/java/com/github/quadflask/smartcrop/SmartCrop.java

```diff
@@ -9,284 +9,375 @@
+        int[] od = o.getRGB();
+        int w = i.width;
+        int h = i.height;
+        int p;
+        int lightness;
+
+        for (int y = 0; y < h; y++) {
+            for (int x = 0; x < w; x++) {
+                p = y * w + x;
+                if (x == 0 || x >= w - 1 || y == 0 || y >= h - 1) {
+                    lightness = 0;
+                } else {
+                    lightness = cd[p] * 8
+                            - cd[p - w - 1]
+                            - cd[p - w]
+                            - cd[p - w + 1]
+                            - cd[p - 1]
+                            - cd[p + 1]
+                            - cd[p + w - 1]
+                            - cd[p + w]
+                            - cd[p + w + 1]
+                    ;
+                }
+
+                od[p] = clamp(lightness) << 8 | (od[p] & 0xffff00ff);
+            }
+        }
+    }
+
+    private void skinDetect(Image i, Image o) {
+        int[] id = i.getRGB();
+        int[] od = o.getRGB();
+        int w = i.width;
+        int h = i.height;
+        float invSkinThreshold = 255f / (1 - options.getSkinThreshold());
+
+        for (int y = 0; y < h; y++) {
+            for (int x = 0; x < w; x++) {
+                int p = y * w + x;
+                float lightness = cd[p] / 255f;
+                float skin = calcSkinColor(id[p]);
+                if (skin > options.getSkinThreshold() && lightness >= options.getSkinBrightnessMin() && lightness <= options.getSkinBrightnessMax()) {
+                    od[p] = ((Math.round((skin - options.getSkinThreshold()) * invSkinThreshold)) & 0xff) << 16 | (od[p] & 0xff00ffff);
+                } else {
+                    od[p] &= 0xff00ffff;
+                }
+            }
+        }
+    }
+
+    private void saturationDetect(Image i, Image o) {
+        int[] id = i.getRGB();
+        int[] od = o.getRGB();
+        int w = i.width;
+        int h = i.height;
+        float invSaturationThreshold = 255f / (1 - options.getSaturationThreshold());
+
+        for (int y = 0; y < h; y++) {
+            for (int x = 0; x < w; x++) {
+                int p = y * w + x;
+                float lightness = cd[p] / 255f;
+                float sat = saturation(id[p]);
+                if (sat > options.getSaturationThreshold() && lightness >= options.getSaturationBrightnessMin() && lightness <= options.getSaturationBrightnessMax()) {
+                    od[p] = (Math.round((sat - options.getSaturationThreshold()) * invSaturationThreshold) & 0xff) | (od[p] & 0xffffff00);
+                } else {
+                    od[p] &= 0xffffff00;
+                }
+            }
+        }
+    }
+
+
+    private void applyBoosts(Image o) {
+        if (options.getBoost() == null || options.getBoost().length < 1) return;
+        int[] od = o.data;
+        for (int i = 0; i < o.width; i++) {
+//            od[i] = (0 << 24) | (od[i] & 0xffffff);
+            od[i] &= 0xffffff00;
```

Comment:
Isn't this going to zero the saturation score? Surely you want to zero the alpha channel (used for boost)?

Reply:
Also, shouldn't you be applying this to the whole image, and not just the first row?

Reply:
> Also, shouldn't you be applying this to the whole image, and not just the first row?

yes, you are right. I fixed that.